src/main/java/com/google/googleidentity/oauth2/validator/AuthorizationEndpointRequestValidator.java

```diff
@@ -0,0 +1,117 @@
+/*
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.validator;
+
+import com.google.googleidentity.oauth2.client.ClientDetails;
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.exception.InvalidRequestException;
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.exception.InvalidScopeException;
+import com.google.googleidentity.oauth2.exception.AccessDeniedException;
+import com.google.googleidentity.oauth2.exception.UnauthorizedClientException;
+import com.google.googleidentity.oauth2.exception.UnsupportedResponseTypeException;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.googleidentity.oauth2.util.OAuth2Utils;
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.Objects;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+public class AuthorizationEndpointRequestValidator {
+
+    private static final Logger log = Logger.getLogger("AuthorizationEndpointParameterValidator");
+
+    public static void validateGET(
+            HttpServletRequest request,
+            ClientDetailsService clientDetailsService) throws OAuth2Exception {
+
+        String clientID = request.getParameter(OAuth2ParameterNames.CLIENT_ID);
+
+        if (clientID == null) {
+            throw new InvalidRequestException("NO_CLIENT_ID");
+        }
+
+        if(!clientDetailsService.getClientByID(clientID).isPresent()){
+            throw new InvalidRequestException("NONEXISTENT_CLIENT_ID");
+        }
+
+        ClientDetails client = clientDetailsService.getClientByID(clientID).get();
```

Comment:
After we switch to storage based ClientDetailsService, here will make two storage queries which will increase the delay. Hence I prefer to keep the outcome of getClientByID in Optional<CLientDetails>, and you can either extract it out into ClientDetails object or keep using the Optional-wrapped instance.